### ISSUE TITLE
RDRAND on AMD CPUs does not work

### ISSUE THREAD

**Issue by hridampalwe** [systemd/systemd#11810]
After suspending the system once can't suspend again. Shutdown gives unmount failed errors and can't shutdown too.
System info 
` System:    Host: Capsparrow Kernel: 4.20.10-1-MANJARO x86_64 bits: 64 compiler: gcc v: 8.2.1 
           Desktop: KDE Plasma 5.15.0 Distro: Manjaro Linux 
  Machine:   Type: Laptop System: HP product: HP 245 G4 Notebook PC v: Type1ProductConfigId 
           serial: <filter> 
           Mobo: HP model: 80C7 v: KBC Version 98.0E serial: <filter> UEFI: Insyde v: F.1C 
           date: 10/29/2015 
Battery:   ID-1: BAT0 charge: 38.8 Wh condition: 39.7/39.7 Wh (100%) model: Hewlett-Packard Primary 
           status: Charging 
CPU:       Topology: Quad Core model: AMD A8-7410 APU with AMD Radeon R5 Graphics bits: 64 type: MCP 
           arch: Puma rev: 1 L2 cache: 2048 KiB 
           flags: lm nx pae sse sse2 sse3 sse4_1 sse4_2 sse4a ssse3 svm bogomips: 17572 
           Speed: 998 MHz min/max: 1000/2200 MHz Core speeds (MHz): 1: 1006 2: 1087 3: 1177 4: 1288 
Graphics:  Device-1: AMD Mullins [Radeon R4/R5 Graphics] vendor: Hewlett-Packard driver: radeon 
           v: kernel bus ID: 00:01.0 
           Display: x11 server: X.Org 1.20.3 driver: ati,radeon unloaded: modesetting 
           resolution: 1366x768~60Hz 
           OpenGL: renderer: AMD MULLINS (DRM 2.50.0 4.20.10-1-MANJARO LLVM 7.0.1) 
           v: 4.5 Mesa 18.3.3 direct render: Yes 
Audio:     Device-1: AMD Kabini HDMI/DP Audio vendor: Hewlett-Packard driver: snd_hda_intel 
           v: kernel bus ID: 00:01.1 
           Device-2: AMD FCH Azalia vendor: Hewlett-Packard driver: snd_hda_intel v: kernel 
           bus ID: 00:14.2 
           Sound Server: ALSA v: k4.20.10-1-MANJARO 
Network:   Device-1: Realtek RTL8111/8168/8411 PCI Express Gigabit Ethernet vendor: Hewlett-Packard 
           driver: r8168 v: 8.045.08-NAPI port: 2000 bus ID: 01:00.0 
           IF: eno1 state: down mac: <filter> 
           Device-2: Broadcom and subsidiaries BCM43142 802.11b/g/n vendor: Hewlett-Packard 
           driver: wl v: kernel port: 2000 bus ID: 05:00.0
Drives:    Local Storage: total: 465.76 GiB used: 68.75 GiB (14.8%) 
           ID-1: /dev/sda vendor: Toshiba model: MQ01ABF050 size: 465.76 GiB 
Partition: ID-1: / size: 147.39 GiB used: 14.49 GiB (9.8%) fs: ext4 dev: /dev/sda1 
           ID-2: /home size: 294.29 GiB used: 54.26 GiB (18.4%) fs: ext4 dev: /dev/sda2 
           ID-3: swap-1 size: 4.00 GiB used: 0 KiB (0.0%) fs: swap dev: /dev/sda3 
Sensors:   System Temperatures: cpu: 52.9 C mobo: 0.0 C gpu: radeon temp: 54 C 
           Fan Speeds (RPM): N/A 
Info:      Processes: 176 Uptime: 1h 07m Memory: 3.32 GiB used: 1.74 GiB (52.3%) Init: systemd 
           Compilers: gcc: 8.2.1 Shell: zsh v: 5.7.1 inxi: 3.0.30 
`
Journalctl : 

`➜  ~ journalctl -xe -p3 -b

-- Subject: A start job for unit network-suspend.service has failed
-- Defined-By: systemd
-- Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- A start job for unit network-suspend.service has finished with a failure.
-- 
-- The job identifier is 1544 and the job result is failed.
Feb 22 20:06:49 Capsparrow systemd[1]: sleep.target: Failed to set invocation ID for unit: File exists
Feb 22 20:06:49 Capsparrow systemd[1]: Failed to start Sleep.
-- Subject: A start job for unit sleep.target has failed
-- Defined-By: systemd
-- Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- A start job for unit sleep.target has finished with a failure.
-- 
-- The job identifier is 1543 and the job result is failed.
Feb 22 20:06:49 Capsparrow systemd[1]: network-resume.service: Failed to set invocation ID for unit: Fi>
Feb 22 20:06:49 Capsparrow systemd[1]: Failed to start Network resume service.
-- Subject: A start job for unit network-resume.service has failed
-- Defined-By: systemd
-- Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- A start job for unit network-resume.service has finished with a failure.
-- 
-- The job identifier is 1628 and the job result is failed.`

**Comment by poettering** [comment]
These failures almost surely are kernel or driver issues. Please contact your downstream distro about this first, and let them escalate issues to us if they are sure this is a systemd issue, which however I think is unlikely.

Also, when you file a bug here, please fill in the form supplied, i.e. provide the systemd version and such. We put that form up for a reason. Thank you for understanding.

**Comment by bl33pbl0p** [comment]
@poettering Someone has reported this in #systemd IRC before. The key thing here is that after the first sleep, systemd returns -EEXIST for unit_set_invocation_id (which uses hashmap operations entirely). 

It might still not be a systemd bug, but the upshot is units end up failing due to this error, so neither does suspend work for the reporter, nor shutdown, in particular, their logs said that multiple jobs failed due to the "Failed to set invocation ID for unit: File exists".

**Comment by bl33pbl0p** [comment]
@poettering Some bug reports in other distributions which were filed very recently:

https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=921267

https://bbs.archlinux.org/viewtopic.php?id=244399

Looks it broke after upgrading to a newer systemd for them...

**Comment by poettering** [comment]
ok

**Comment by Experimenter** [comment]
Facing a similar issue, downgrading systemd to v239 solves the problem. The same reappears on v240 & v241. Tested with Manjaro and Debian sid.

**Comment by TheBeasT15** [comment]
@Experimenter Yes I did that now its resolved.

**Comment by sandy-8925** [comment]
I'm also facing this problem all the time, on a fresh Arch Linux install, on a laptop with an AMD CPU.

I have another laptop and desktop, each with Intel CPUs and neither face this problem. Any chance that this specifically happens on AMD CPUs?

**Comment by madhur4127** [comment]
As mentioned by @sandy-8925, I am also facing this problem on AMD CPU. 

I can't suspend after suspending one time and also cannot shutdown once I have suspended. This problem doesn't happen on shutdown when I don't shutdown once.

`journalctl -r` before reboot shows this:
```
Mar 16 22:25:37 madhur4127 systemd[850]: systemd-exit.service: Job systemd-exit.service/start failed with result 'dependency'.
Mar 16 22:25:37 madhur4127 systemd[850]: exit.target: Job exit.target/start failed with result 'dependency'.
Mar 16 22:25:37 madhur4127 systemd[850]: Dependency failed for Exit the Session.
Mar 16 22:25:37 madhur4127 systemd[850]: Dependency failed for Exit the Session.
Mar 16 22:25:37 madhur4127 systemd[850]: Failed to start Shutdown.
Mar 16 22:25:37 madhur4127 systemd[850]: shutdown.target: Failed to set invocation ID for unit: File exists
```

**Comment by TheBeasT15** [comment]
Yep I am on AMD too I tried on Intel CPU worked flawlessly.

**Comment by Experimenter** [comment]
For me too only the AMD CPUs have been effected. My machines with AMD A8 
& A6 have a problem while the intel one runs fine. After suspend every 
service I try to startup using systemctl pops up the same error: Failed 
to setinvocation IDforunit: Fileexists

**Comment by thodnev** [comment]
Also got AMD CPU and experienced the same problem after latest Arch upgrade. 
Downgraded packages `systemd`, `lib32-systemd` and `systemd-sysvcompat` to version 239.6-1 and it started working like a charm.
The problem is that it got packaged and distributed to many users as stable update.
Probably distro maintainers should be more careful.
[sudo lshw](https://github.com/systemd/systemd/files/3016484/lshw.log)
[journalctl -b -1](https://github.com/systemd/systemd/files/3016485/boot.log)

**Comment by oyvinds** [comment]
got this issue on Fedora 30 with kernel 5.0.6 on "AMD E1-6010 APU with AMD Radeon R2 Graphics". systemd version systemd-241-4.gitcbf14c9.fc30.x86_64.

After suspend any and all services fail to start with "Failed to set invocation ID for unit: File exists" and the machine is essentially unusable. 

Tried using both radeon and amdgpu for the gpu part of the machines apu, that makes no difference. 

I'm not sure where to begin figuring out why systemd breaks after return from suspend. Let me know if/how I can provide more information.

**Comment by ChrisJAllan** [comment]
I have this issue on my laptop with "AMD A6-6310 APU with AMD Radeon R4 Graphics", but not on my desktop with an AMD CPU and an nVidia GPU, so it's probably something on the GPU end.

**Comment by oyvinds** [comment]
Just a little detail, suspend to disk doesn't have this problem on my machine - just suspend to RAM. Thus, using suspend to disk instead is a somewhat acceptable workaround.

**Comment by ranjan-purbey** [comment]
In my case, the problem occurs not only with suspend but also after restarting any of the systemd services. For e.g., running the following:
```
$ sudo service systemd-resolved restart
```
causes any subsequent attempts to start/restart a service to fail.
Yes, I am also on AMD CPU + Radeon GPU

**Comment by ranjan-purbey** [comment]
Even v142 has the same issue. So for the time being, downgrading to v139 seems to be the only work-around.

**Comment by mbiebl** [comment]
Apparently, this issue is easy to trigger, if you have the right hardware (and unfortunately none of the developers seem to have those).
So, if anyone who is affected by this issue can run a `git bisect` between v239 and v240 to find the first faulty commit, this would be super helpful.

**Comment by oyvinds** [comment]
Triggering this really is super-easy on the AMD E1-6010, it's one of those happens-every-time bugs. Doing a git bisect on a 1.4 GHz dual-core isn't very tempting, though. That's the only hardware I have where this happens, Ryzen+RX570 doesn't have this problem. I could do if nobody else with hardware that doesn't spend days compiling systemd (or anything) does it.

**Comment by mbiebl** [comment]
@oyvinds That would be great, thanks! If the AMD E1-6010 is indeed that slow, I would consider compiling it on your faster Ryzen system and copying the binaries/rpm over.

**Comment by madhur4127** [comment]
@oyvinds @mbiebl, I am also affected by the bug and I want to help, but I am new to Open Source and require guidance as this project is so big!

I have AMD A8-7410 quad-core 2.2Ghz, so I think it would suffice.

**Comment by sandy-8925** [comment]
I haven't used that problematic system recently, will see if I can still reproduce, and if yes then I will try bisecting.

**Comment by jimy-byerley** [comment]
Hello everyone.
I have the same problem with an AMD E1 and AMD radeon R2. My laptop can't suspend a second time and then can neither shutdown.
I got the same message in `journalctl`:
```
mai 04 12:09:20 wopr systemd[1]: systemd-suspend.service: Failed to set invocation ID for unit: File exists
mai 04 12:09:20 wopr systemd[1]: Failed to start Suspend.
-- Subject: L'unité (unit) systemd-suspend.service a échoué
-- Defined-By: systemd
-- Support: https://www.debian.org/support
-- 
-- L'unité (unit) systemd-suspend.service a échoué, avec le résultat failed.
```

**Comment by mbiebl** [comment]
I guess at this point we don't need further "me too's", as this doesn't really help us to find the root cause. Instead we need someone with the appropriate hardware to find the first faulty commit.

**Comment by jimy-byerley** [comment]
I was trying to make a git bisect on my side, but at every step, the "./configure" step fails with the following error:
```
Meson encountered an error in file /home/jimy/systemd/meson_options.txt, line 171, column 53:
Expecting rparen got id.
```
What does this message mean ?

**Comment by madhur4127** [comment]
@mbiebl, I am currently bisecting between v239 and v240. 

Using `systemd-nspawn -bi image.raw` to test images causes image to be loaded in terminal itself. How can I test whether the current commit is good/bad?

I thought of using `systemctl suspend` twice to test for image but the error displayed is: `Failed to suspend system via logind: Sleep verb "suspend" not supported`

**Comment by mbiebl** [comment]
@madhur4127 You'll need to run systemd no bare-metal and not inside a systemd-nspawn container. You can't suspend a container via `systemctl suspend`.

**Comment by jimy-byerley** [comment]
I was trying to compile systemd to do a git bisect, but I went to an error during `./configure` because my version of `libmount` (use by command mount) is to old (<= 2.30, I use a debian stable for the compilation)
I tried to recompile libmount myself, from util-linux, but after that `pkgconfig`  continues to say my libmount version is 2.29. Is there a way to correct it ?

**Comment by vcaputo** [comment]
@jimy-byerley In my experience it's simpler to just modify the meson.build file where the libmount >= 2.30 dependency is specified.  However Debian stable's version of Meson is also too old for the systemd meson.build syntax, so I also have a newer Meson version built from source in /usr/local.

With those two changes systemd builds on Debian stable for me, unless I'm forgetting other things it needed.

**Comment by wtdrog** [comment]
I think I have found a reason why this issue happens. Commit https://github.com/systemd/systemd/commit/cc83d5197ca08d68fa78167b6a64e9f28da3cc96 introduced optional randomness generation based on RDRAND instruction. So, if RDRAND instruction is present, it is used to generate UUIDs for invocation IDs. There is a bug mentioning problems with this instruction after resuming on AMD CPUs here: https://bugzilla.redhat.com/show_bug.cgi?id=1150286.
I have a laptop with AMD A6-6310 CPU, so this problem affects my CPU. I upgraded my Fedora installation to 30th release yesterday and have met this bug while trying to suspend my laptop this morning. The installed version of systemd was 241-8.git9ef65cb.fc30. I downloaded RPM sources and changed line `r = genuine_random_bytes(&t, sizeof t, RANDOM_ALLOW_RDRAND);` in function `sd_id128_randomize` in `src/libsystemd/sd-id128/sd-id128.c` to not use RDRAND at all while generating UUIDs (`r = genuine_random_bytes(&t, sizeof t, 0);`). After the package was recompiled I checked that I use exactly the same version as I've patched, then rebooted, tried to sleep several times, failed, installed patched package, rebooted, tried again to sleep several times, now successfully.

**Comment by ChrisJAllan** [comment]
Confirming the above fix. Tested on systemd 242.19 on Arch with an A6-6310.

**Comment by poettering** [comment]
@wtdrog so that sounds as if RDRAND returns non-random data after suspend on these AMD systems? yikes. That sound like a hardware/firmware/kernel bug and a major security problem with these CPUs...

https://bugzilla.kernel.org/show_bug.cgi?id=85911

Given that many plenty distros (including Fedora) nowadays seed the kernel entropy pool with RDRAND this is really something the kernel folks should look into. This looks like something you really want to involve the relevant kernel people in...

**Comment by mbiebl** [comment]
Is there a way to detect such a broken RDRAND implementation and fall back to `getrandom()` automatically? This feels like a rather nasty issue and I wonder what to do about that in the upcoming Debian release which will ship with v241.

**Comment by poettering** [comment]
@mbiebl this looks like something better fixed or worked around in the kernel.

**Comment by marksteward** [comment]
The kernel only xors in rdrand as an additional entropy source, not the only one.

See https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=c2557a3 and https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=85a1f77. Seeding is vulnerable to a malicious rdrand, but not a buggy one, assuming that's what this is.

**Comment by mbiebl** [comment]
> @mbiebl this looks like something better fixed or worked around in the kernel.

@bwhacks can you chime in here. Is it realistic that a fix/workaround for this in the kernel will land for buster?

**Comment by mbiebl** [comment]
fwiw, the alternative I would consider is to disable the usage of RDRAND in systemd until such a kernel fix exists.

**Comment by bwhacks** [comment]
@mbiebl It seems like the proper thing to do in the kernel would be to mask out the RDRAND feature on processors where it's unreliable. Of course you can mitigate this in systemd too so you don't depend on that getting upstream and into all the different kernel branches people may use.

**Comment by bwhacks** [comment]
But I'm not sure how we would check for this in the kernel at boot time, other than with a blacklist of DMI identifiers. Blacklisting all AMD processors seems like too extreme a response.

Given that RDRAND is allowed to fail, it seems to me that you should either try it only once, or only a few times, before falling back to whatever code is used when RDRAND is not implemented.

**Comment by nicowilliams** [comment]
@bwhacks Don't check for this, just don't use only `RDRAND`.  Instead use a proper, seeded PRNG, and if you really really want `RDRAND` then XOR `RDRAND` with the output of the PRNG.

**Comment by edef1c** [comment]
@nicowilliams At that point we should just be using getrandom for seeding the PRNG, and stop caring about RDRAND in userspace entirely.

**Comment by nicowilliams** [comment]
@edef1c: perfect, do that.

**Comment by wtdrog** [comment]
I have checked values generated by RDRAND on my CPU: after resuming from suspend it is always -1 (or 0xFFFFFFFFFFFFFFFF). I think that on other buggy CPUs that value should be the same. So if it is, it would be possible to not remove RDRAND entirely, but instead add another check for the generated value.

**Comment by nicowilliams** [comment]
Folks, what's wrong with using `getrandom()`?

Using only `RDRAND` directly feels wrong precisely because we shouldn't trust HE RNGs this much.  It feels like you're using just to be cool.  "Look 'ma!  No syscalls here!"?

**Comment by ignaciocaamanio** [comment]
Same issue with an AMD A6-6310 APU with AMD Radeon R4 Graphics.

**Comment by howaboutsynergy** [comment]
> I have checked values generated by RDRAND on my CPU: after resuming from suspend it is always -1 (or 0xFFFFFFFFFFFFFFFF). I think that on other buggy CPUs that value should be the same. So if it is, it would be possible to not remove RDRAND entirely, but instead add another check for the generated value.

Hi. Would you please consider sharing the source code for doing that? 
I'd like to check on `AMD A6-3400M APU with Radeon(tm) HD Graphics`

**Comment by valarauca** [comment]
@wtdrog Is the CF flag also set when `-1` is returned? 

I am assuming it is, but when CF is not set, it implies the returned value is invalid and the processor cannot return data at this time.

**Comment by wtdrog** [comment]
@valarauca Exactly. RDRAND always sets CF.

**Comment by poettering** [comment]
> The kernel only xors in rdrand as an additional entropy source, not the only one.

It recently optionally credits entropy to RDRAND, and on many distros (such as Fedora) that is enabled by default. And if RDRAND doesn't work, and you credit it for entropy then you have a problem.

**Comment by poettering** [comment]
> Is there a way to detect such a broken RDRAND implementation and fall back to `getrandom()` automatically?

I guess somebody with an AMD CPU needs to look into that...

**Comment by poettering** [comment]
> fwiw, the alternative I would consider is to disable the usage of RDRAND in systemd until such a kernel fix exists.

then you punish everyone on intel... we can certainly mask this out for amd cpus but if you ask me this sounds like something the kernel sould deal with, not us.

**Comment by poettering** [comment]
> Given that RDRAND is allowed to fail, it seems to me that you should either try it only once, or only a few times, before falling back to whatever code is used when RDRAND is not implemented.

We try it once, check if it failed, and if it did use other sources of randomness. Note we only use RDRAND for stuff where the highest quality of entropy isn't required. i.e. we use it for UUID generation, not for crypto keys and such.

**Comment by poettering** [comment]
BTW, the reason we use RDRAND instead of getrandom is that we need to generate uuids early on (since every service we starts gets one passed, the "invocation ID", and for other stuff too), but getrandom complains in dmesg or blocks if we call it before the pool is initialized. To avoid that we use RDRAND when generating uuids, since it should be good enough for that, as the usecase needs a "mid-quality" rng source: not crypt quality and not totally guessable either.

**Comment by tytso** [comment]
If systemd can generate a per-system secret key, and can store it someplace where only root can gain access, and can assume that its clock is accurate (and isn't always getting reset to January 1, 1970, or something silly like that), then one thing you can do is encrypt the time by the per-system key, and that might be good enough for your rng source.  These assumptions might not be true for all systems that systemd might want to target, but maybe it can be one approach that will work well enough for many laptop/desktop/server systems?

**Comment by mbiebl** [comment]
> then you punish everyone on intel... we can certainly mask this out for amd cpus but if you ask me this sounds like something the kernel sould deal with, not us.

I agree with you, that ultimately the kernel should deal with this. That said, I think we should also address this from the systemd side for the time being, until a fixed kernel is available.
The current failure mode on affected systems is simply too nasty to not deal with this.

**Comment by poettering** [comment]
@tytso generi,c classic linux OSes are usually booted with an initrd, and those then search for the storage to use. These initrds run systemd themselves (as finding the storage might involve plenty services to make complex storage work). This means in order to find the secret key you need the storage, but to find the storage we start plenty services and thus want to generate uuids and thus need the secret key... So you have an ordering cycle here. 

The good thing about RDRAND is that when available is available without any pre-condition of having seen storage already.

@tytso btw, is there a way how the boot loader could supply some random seed to the kernel to seed its initial pool from? i.e. does the kernel maybe accept a kernel cmdlne option with random data? if so it should be easy to teach a boot loader to look for maybe a special raw partition that only contains random seed data, read that, mark it as invalidated and pass it to the kernel via the kernel cmdline. During OS boot the partition could then be filled with new seed data. That way, the kernel would always come up with a full pool, and the "saved random seed" logic would not be something we do during late boot (i.e. too late to be useful).

**Comment by mbiebl** [comment]
Btw, is using `getrandom()` still problematic during early boot if the kernel has been built with `CONFIG_RANDOM_TRUST_CPU=y`?

**Comment by poettering** [comment]
@mbiebl does debian set CONFIG_RANDOM_TRUST_CPU=y? I think doing that is generally a good idea. However we probably still want to support either case in systemd. (And dunno, maybe there's also the theoretical chance that CONFIG_RANDOM_TRUST_CPU stuff gets confused by RDRAND failing if you can make the system suspend so early that the RDRAND feeding thread hasn't completed its work yet...)

**Comment by mbiebl** [comment]
@poettering yes, the Deban kernel does set `CONFIG_RANDOM_TRUST_CPU=y`.
Which is why I was considering reverting cc83d5197ca08d68fa78167b6a64e9f28da3cc96 for Debian (at least for buster). Or is there still a benefit of using RDRAND directly?

**Comment by poettering** [comment]
@mbiebl i think if you set that kernel option getrandom() should generally be fine to always use on systems with RDRAND. But I figure @tytso might know this better. IIUC the kernel will feed RDRAND data into the pool until the pool is initialized, inside a kernel thread specific to that purpose. I would assume that thread is finished with everything by the time userspace is invoked, but not sure if that's guaranteed. If it isn't htere might still be value in doing RDRAND from userspace here even on kernels that do have CONFIG_RANDOM_TRUST_CPU=y set, because otherwise systemd would race against that feeder thread...

**Comment by tytso** [comment]
@poettering Sure, the right long-term answer is to make this to be the bootloader's problem.   We could extend the boot loader protocol so that in addition to the initrd and boot command line, the boot loader could also pass to the kernel a 32 byte "secure random seed".   That just pushes the problem back one level, and raises the question how much do *you* trust bootloader authors?   Or for that matter, if the bootloader is going to be asking for random numbers from the UEFI BIOS, BIOS authors?    They probably have the same level of competence as the people who wrote Intel's Management Engine (IME) and look what a disaster *that* was.   Worse, the UEFI BIOS is all closed source, and not auditable.    Still, this is the solution that OpenBSD chose, and given that we assume that many bootloaders have access to file systems (although in some cases the quality of that code is also not one that has made, say, the XFS developers very happy), it does mean that we can assume the bootloader can read a secure seed file from some file in the root file system, and then hopefully we can coordinate with systemd or other userspace components to refresh the secure seed file (a) as part of the initial bootup, and (b) also as part of the clean shutdown procedure.   The problem is actually implementing this solution.   Getting people with both (a) skills, (b) time, and (c) access directly, or indirectly, to get changes pushed into all of the various components, has just not happened up until now.    We basically need people to implement an example with patches to grub, the kernel, and some systemd unit files.   Maybe a GSOC project?

As far as how getrandom(2) works, we always will mix in RDRAND if it is available.  This is separate from using RDRAND to help initialize the CRNG.    Without CONFIG_RANDOM_TRUST_CPU, we'll mix in RDRAND since it can never make things worse, and hopefully will make things better ---- but we don't give any *credit* for the contributions from RDRAND, either.   This is essentially to satisfy those people who are convinced that Intel let the NSA insert a back door into RDRAND, and/or looked at the quality of Intel's Management Engine implementation, and decided not to trust RDRAND on grounds of competence, as opposed to theories of secret collaboration with the US government.   Even if you think that RDRAND is massively incompetently written, such that it's always returning 0xFFFFFFFF, XOR'ing a constant isn't going to make things any worse.

Ultimately, whether or not you enable CONFIG_RANDOM_TRUST_CPU is really more of a political question than anything else.   Personally, I figure that if you are willing to use Intel CPU's, you had better darn well trust them.   And if you don't trust them to get RDRAND right, how would you be trusting them not make their CPU massively susceptible to side-channel cache timing attacks?   (Oh, wait....  :-)

BTW, programs using RDRAND are supposed to check the Carry Flag (CF); if the CF is 0, that means no random value was available.    So if AMD is returning a failure after a suspend resume, that's an unfortunate implementation, but it's not an outright violation of the abstraction contract.   If it is returning a CF of 1, *and * always returning 0xFFFFFFFF, that's a complete disaster, and customers should be screaming at AMD for a refund.    Was systemd checking the CF flag after calling RDRAND?

**Comment by oyvinds** [comment]
> Is there a way to detect such a broken RDRAND implementation and fall back to `getrandom()` automatically? 
You could check when RDRAND is used. Get a value from RDRAND and if it's 0xFFFFFFFFFFFFFFFF you call back to getrandom().

You could also check if the CPU is AMD CPU family 22. Family 21 doesn't have RDRAND and Family 23 (Ryzen) is fine.

**Comment by poettering** [comment]
> @poettering Sure, the right long-term answer is to make this to be the bootloader's problem. We could extend the boot loader protocol so that in addition to the initrd and boot command line, the boot loader could also pass to the kernel a 32 byte "secure random seed". That just pushes the problem back one level, and raises the question how much do _you_ trust bootloader authors? 

well, there's sd-boot (a uefi boot loader we came up with in the systemd context), whose authors I tend to trust... ;-)

> Or for that matter, if the bootloader is going to be asking for random numbers from the UEFI BIOS, BIOS authors? They probably have the same level of competence as the people who wrote Intel's Management Engine (IME) and look what a disaster _that_ was. Worse, the UEFI BIOS is all closed source, and not auditable. Still, this is the solution that OpenBSD chose, and given that we assume that many bootloaders have access to file systems (although in some cases the quality of that code is also not one that has made, say, the XFS developers very happy), it does mean that we can assume the bootloader can read a secure seed file from some file in the root file system, and then hopefully we can coordinate with systemd or other userspace components to refresh the secure seed file (a) as part of the initial bootup, and (b) also as part of the clean shutdown procedure. The problem is actually implementing this solution. Getting people with both (a) skills, (b) time, and (c) access directly, or indirectly, to get changes pushed into all of the various components, has just not happened up until now. We basically need people to implement an example with patches to grub, the kernel, and some systemd unit files. Maybe a GSOC project?

I am certainly willing to work on the systemd side of things. I figure the kernel changes would be small I presume (in particular if this is implemented via a new kernel cmdline arg). Grub otoh...

BTW, I know of some cloud people who are interested in provisioning each image they spawn in their environments with an individually generated random seed. For cases like that it would be perfect if the 32byte value you suggested could be supplied via the kernel cmdline, since they have nice deployment infrastructure for controlling that already, and it needs no boot loader patching. Hence I'd actually prefer supplying the random seed via cmdline arg rather than boot protocol structures.
 
> As far as how getrandom(2) works, we always will mix in RDRAND if it is available. This is separate from using RDRAND to help initialize the CRNG. Without CONFIG_RANDOM_TRUST_CPU, we'll mix in RDRAND since it can never make things worse, and hopefully will make things better ---- but we don't give any _credit_ for the contributions from RDRAND, either. This is essentially to satisfy those people who are convinced that Intel let the NSA insert a back door into RDRAND, and/or looked at the quality of Intel's Management Engine implementation, and decided not to trust RDRAND on grounds of competence, as opposed to theories of secret collaboration with the US government. Even if you think that RDRAND is massively incompetently written, such that it's always returning 0xFFFFFFFF, XOR'ing a constant isn't going to make things any worse.

BTW, could we please get CONFIG_RANDOM_TRUST_TPM= as well? I am pretty sure people who have a TPM would trust it as much or as little as the CPU itself...

> BTW, programs using RDRAND are supposed to check the Carry Flag (CF); if the CF is 0, that means no random value was available. So if AMD is returning a failure after a suspend resume, that's an unfortunate implementation, but it's not an outright violation of the abstraction contract. If it is returning a CF of 1, *and * always returning 0xFFFFFFFF, that's a complete disaster, and customers should be screaming at AMD for a refund. Was systemd checking the CF flag after calling RDRAND?

Yes, it appears that AMD is not correctly setting CF. We check for that, and are actually entirely fine if it's not set and will fall back to getrandom() in that case. 

See: https://github.com/systemd/systemd/blob/master/src/basic/random-util.c#L54

**Comment by poettering** [comment]
> You could also check if the CPU is AMD CPU family 22. Family 21 doesn't have RDRAND and Family 23 (Ryzen) is fine.

is that verified? i presume cpuid will tell me the family? do you have example code maybe?

**Comment by poettering** [comment]
i presume i would first call cpuid with eax=0 and check for AuthenticAMD, and then call it with eax=1 and check for the family bits?

**Comment by poettering** [comment]
that said i wonder though if it wouldn't be easier to simply refuse accepting ULONG_MAX as random nr, and if we see it consider that equivalent to cf == 0. And do this regardless of the CPU vendor. However, for that it would be good to know if the affected cpus really always return ULONG_MAX and nothing else.

**Comment by poettering** [comment]
hmm, so here's a patch doing the ULONG_MAX check:

https://paste.fedoraproject.org/paste/Qhao0f9NszPj8K9EgCSbnw

Maybe people with an affected system could give that a whirl? it's a very simple way to hopefully filter out the bogus cases, under the assumption rdrand always returns ULONG_MAX when it fails like this.

**Comment by dj-on-github** [comment]
Returning RdRand-->{c=0,result=-1} on an RNG underflow is inconsistent with the RdRand instruction spec. On underflow the carry is cleared and 0 is returned. This behaviour can be seen with RdSeed if you pull fast enough. The Intel SDG says to retry RdRand (based on carry) up to 10 times and then infer a failure if you don't get cc=1 by then (thus filtering hypothetical transient underflows). I don't work for AMD so I can't intuit what's going on, but where I work, lots of engineering went into meeting the resume dwell time so that this kind of problem doesn't happen on resume.

A note of caution on the TPM - If you're using the TPM RNG as the only source of entropy and you're trying to do the secure transaction (I forget the name) on the TPM wires, there is no entropy on the host side for the nonce. You're literally using the RNG within the TPM to provide the entropy for the nonce for both the TPM and the Host messages. A security mess if ever there was one, especially with a MITM between the TPM and host.

**Comment by poettering** [comment]
> @poettering would it make sense to log about it if we get crap data (say with `log_debug()`)?

Sure, we can do that, if we decide to merge the patch. In its current form it's only an excercise though, to see if this actually works around the issue. Would be great if anyone affected by this could test it and see if it fixes things for them properly. If so, we can rework this and merge it. If noone is willing to test this this will remain open though, sorry.

**Comment by jimy-byerley** [comment]
I can test, I have a fast processor to compile for my small one. If the trick of @vcaputo of changing the libmount version in systemd dependencies is working at least ..

**Comment by mbiebl** [comment]
I've compiled Debian packages for sid/buster. In case anyone running Debian wants to test the patch from @poettering 
See https://people.debian.org/~biebl/systemd/buster/

**Comment by jimy-byerley** [comment]
Suspending and resuming after is working again thanks to the packages of @mbiebl !
I tried 4 times to close/open my laptop and it works well again.
Is there any journalctl information you need ?

**Comment by mbiebl** [comment]
thanks for testing @jimy-byerley 
For completeness sake, can you attach your lscpu output.

**Comment by poettering** [comment]
I prepped a version of the earlier patch that should be good enough to commit now in #12536. PTAL.

Note that this is really just an ugly work-around, and we really shouldn't have to do this in userspace. Somebody who cares about this really should look into this, and put some pressure behind getting this fixed properly, i.e. in the Linux kernel: the CPU feature should probably be masked if RDRAND doesn't really work on those faulty CPUs. Maybe that someone should also contact AMD technical folks about this. It's unlikely that systemd is the only userspace tool affected by this, and the linux kernel might be as well if it credits entropy to an RDRAND implementation that has no entropy...

I am not going to be that someone though, my interest in AMD CPUs is relatively limited.

**Comment by mbiebl** [comment]
@tytso do you know someone from AMD who could chime in here?

**Comment by Hunman** [comment]
@poettering I have an AMD A4 with this issue, I can try that workaround after work

**Comment by mamins1376** [comment]
E1-6010 owner here; This affects me too. And by the way the patch provided by @poettering is gone.

**Comment by ignaciocaamanio** [comment]
I have this issue with my AMD cpu and wanted to test this, but I can't build systemd. I use archlinux, mkosi fails with a lot of "target not found errors: <some-pkg-devel>". Those pkg aren't present in arch, I don't know how to fix it.

**Comment by arianvp** [comment]
Yo @ignaciocaamanio I would suggest taking the `PKGBUILD` file of the systemd package in the arch Linux source code and modifying it to point to the patch.  Then you can utilise arch's build system to build a custom version of the package and it will install all the required dependencies for you

**Comment by ignaciocaamanio** [comment]
I coulded build systemd with `asp checkout systemd` and then `meson build` and `ninja -C build` but fails at `sudo mkosi`, tries to install a lot of missing pkg-devel packages. I'm going to try your suggestion @arianvp

**Comment by yrahul3910** [comment]
@poettering I'm on Fedora 30 with an affected system (AMD A8 APU). It seems the pastebin link is broken, could you upload a new one? This is a rather serious bug. My `lscpu` output shows that I do have `rdrand`.

**Comment by yrahul3910** [comment]
@poettering I also seem to have this bug (I'm on the KDE Spin), where after waking up from suspend, and typing in the password, there's a 20 second delay before the desktop is shown. Could this bug be causing this issue as well?

**Comment by forthy42** [comment]
If you care about entropy, and you have access to the file system, this can help:

Have a file that contains randomness initialized at install time.
At boot, add that file's contents to the entropy pool, using RNDADDENTROPY.
Read /dev/urandom and overwrite that file with new randomness.
To be sure you won't do this twice in a row, open the file, delete it, then read it, recreate it, write it.
If the file is either empty or not available, you can only recreate it, and need to make sure you have the entropy to do so.  When you create it first time or after an error condition, use /dev/random to fill it.

**Comment by poettering** [comment]
@forthy42 we do that anyway, that's what systemd-random-seed is for, a standard component of systemd.

@yrahul3910 the fix got commit, see #12536. ask your distro to backport it to your version.

**Comment by mbiebl** [comment]
I read today, that apparently even the latest AMD CPUs are affected:
https://www.heise.de/newsticker/meldung/Ryzen-3000-Bekannter-Fehler-bringt-viele-Linux-Distributionen-zu-Fall-4465220.html

**Comment by ignaciocaamanio** [comment]
This fix is going to appear in systemd 243 ?

**Comment by ruabmbua** [comment]
Apparently yes.

**Comment by ignaciocaamanio** [comment]
Is there a release date for 243?

**Comment by benbucksch** [comment]
From the patch:
> `asm volatile("rdrand %0;"`
> The Linux kernel should probably work around this issue

I don't understand why you call assembler code directly in the first place. Why not use the Linux functions to get random numbers? There, Linux has the ability to decide which hardware features to use to generate random numbers, including combining several sources of randomness. A user also has the option to distrust hardware random number generation altogether (you see here why). By calling CPU functions per assembler, you're circumventing all that.

**Comment by forthy42** [comment]
@benbucksch RdRand is only used when the Linux kernel says it has not enough entropy yet.

**Comment by benbucksch** [comment]
@forthy42: Thanks.

FYI: Commit 97fa202 introduced this and says why.

**Comment by forthy42** [comment]
BTW: The checked in code is broken. It checks for the “please do rdrand again” condition, returns -EAGAIN if so, and then the caller converts this to a hard error message instead of calling again.

Please, use rdrand as it is meant to:

    rdrand_loop:
        rdrand %0
        jnc rdrand_loop

Anything else will just break.

**Comment by keszybz** [comment]
@forthy42 try reading more carefully.

**Comment by forthy42** [comment]
Some invocations of rdrand() fall back to libc's rand()… instead of calling again.  This sort of code is error-prone; don't do that.

BTW: If all you need is a non-repeating sequence of lousy quality pseudo random 64 bit numbers, read one single seed, feed it into an xorshift PRNG, and use that output for the rest.  It's absolutely guaranteed to not repeat itself, regardless if that first seed is 0 or -1 or really random.

**Comment by mbiebl** [comment]
> Some invocations of rdrand() fall back to libc's rand()… instead of calling again. This sort of code is error-prone; don't do that.
> 
> BTW: If all you need is a non-repeating sequence of lousy quality pseudo random 64 bit numbers, read one single seed, feed it into an xorshift PRNG, and use that output for the rest. It's absolutely guaranteed to not repeat itself, regardless if that first seed is 0 or -1 or really random.

@forthy42 This sounds like a reasonable request. Could you file it as RFE and post your input there?

**Comment by poettering** [comment]
> Some invocations of rdrand() fall back to libc's rand()… instead of calling again. This sort of code is error-prone; don't do that.

Please try to understand the usecases for the fallback-to-rand() logic. It's documented in long detail here:

https://github.com/systemd/systemd/blob/master/src/basic/random-util.c#L36

There's really no point in warming up this discussion over and over again. Instead, please just read the comments, it saves everyone time.

We fall back to rand() if rdrand() is only used where the quality really doesn't matter much (such as: a loop around temporary file generation which includes a random number, where low entropy just means we loop a bit longer, during early boot where the kernel's own pool is not filled yet, at which point we use that.).
 
> BTW: If all you need is a non-repeating sequence of lousy quality pseudo random 64 bit numbers, read one single seed, feed it into an xorshift PRNG, and use that output for the rest. It's absolutely guaranteed to not repeat itself, regardless if that first seed is 0 or -1 or really random.

No, let's not roll our own PRNGs, we don't want to maintain that. it's fine to use the CPU RNG, the kernels or glibc's, but I don't want to maintain our own version in our codebase, sorry. Also, why even? glibc is fine for low quality randomness. If you think glibc's implementation sucks, please work with them to fix that.

> @forthy42 This sounds like a reasonable request. Could you file it as RFE and post your input there?

That's just hiding a bug that is elsewhere. In our own code we handle the case fine too, and log about it, that's already good enough. I mean, if we'd just generate a series froever from a single constant value that means when we ever need low quality entropy (for cases such as temporary file name generation) we are forever guessable, which sucks...

**Comment by forthy42** [comment]
> No, let's not roll our own PRNGs, we don't want to maintain that. it's fine to use the CPU RNG, the kernels or glibc's, but I don't want to maintain our own version in our codebase, sorry. Also, why even? glibc is fine for low quality randomness. If you think glibc's implementation sucks, please work with them to fix that.

It used to be an LCG, and if this is good enough for you, just use it.  glibc has already an improved random() function, and the current version uses that for rand(), too (the LCG is gone for good).  The random() function passes some of those tests where the LCG fails, and has enough internal state for non-cryptographic use-cases.  It's not in a state I would complain about.  Given sufficient seeding, it's good.  RdRand, if correctly implemented, is cryptographically good, so far better.  But you can't rely on that.

The question would be rather why use rdrand() for anything else than seeding rand(), since you can't rely on it.  You need to know it's good to actually prefer it.

All you need is to seed rand(); using rdrand for this purpose is good (as you do now), because it's just once, and if the rdrand seed is deterministic, it only fails security concerns (the attacker then can guess all the init states of the hash tables if it has some clues about the boot time).  Logging the seed value would be a good idea, because people can check if it looks random.  It won't crash if it isn't.

Essentially, this general approach (seed with monitored entropy source+deterministic, well-known PRNG that is never reseeded once it's running) is what DJB also suggests for cryptographic PRNGs:

http://blog.cr.yp.to/20140205-entropy.html

This approach is sane, easy to maintain and won't fail, and problems can be monitored.  If you don't need high quality randomness, then your PRNG can indeed be simple (LCG probably a bit too bad, but random() is good).

Actually, you already do that, if rdrand() fails.  All you need is to replace the for(;;) loop at line 179 with the fallback to pseudo_random_bytes() and not bother with rdrand() outside of initialize_srand().  There, it's only an entropy source, and you can use it (together with at least one other lousy entropy source, monitored).  The only suggestion I would have for pseudo_random_bytes is to use the most significant 24 bits of rand() (>>23, >>15, >>7), not the least, if you compile for an old glibc version where rand() actually still is the LCG.  With rand()=random(), it doesn't matter.

The security researcher problem with rdrand() is that it is a black box, and can fail in unexpected ways.  You can't audit it, and you can't expect it to be reasonable, or that it sets the “I failed” flag when it failed, or that retry is actually a cure (the CF is a “please retry” flag in Intel's documentation).  Ryzen's behavior proves that pessimistic assumption to be absolutely correct.

**Comment by poettering** [comment]
> The question would be rather why use rdrand() for anything else than seeding rand(), since you can't rely on it. You need to know it's good to actually prefer it.

The comments explain that already.

We assign UUIDs for all service invocations, to make them recognizable in logs and stuff so that we can associate logs with their invocations properly. For UUIDs using rand() is not good enough, since the results are likely not going to be unique. OTOH we don't need cryptographic quality either, since these are not key material. We used to pull UUIDs from /dev/urandom, which sucks during early boot though, since the kernel complains if we do that while the pool isn't fully initialized. We can't use getrandom() for this usecase either, since it either fails entirely or blocks during early boot, and possibly for a long time. So we try to avoid using /dev/urandom for such stuff if we can, by using RDRAND, which is not superduper fast, but also not super dupler slow, doesn't result in log spew. When RDRAND doesn't work though we will fall back to /dev/urandom for thus, but never to rand().

In general: we have to trust that CPUs work. If they don't we are fucked. We don't have to trust them to the level to directly derive cryptographic keys from it (though we usually do now, most bigger distros trust RDRAND now for /dev/urandom filling), but we certainly have to trust them that the basics work. And no, on those AMD CPUs the basics don't work at all, apparently. I mean, if you can't even rely on AMD CPUs to give you a non-constant value on RDRAND, what's next? that it returns a constant value on ADD? I mean, we have to put the line somewhere...

And anyway, we detect the constant values these days, and log. It sucks hard, but it does what is needed to to do. If we'd just seed a systemd-private PRNG from RDRAND exactly zero is won, since the AMD CPUs don't fail when you query many RDRAND values, but alread yon the first. So what would you gain, besides having to maintain your own PRNG which we really don't want?

And anyway, random-util.c has pretty much more comments than code these days, explaining all our usecases, with a strong focus on the fact we need all this during early boot. Instead of warming that up again and again, why not just have a look there and read up? Because this stuff *has* been discussed before, and you are not really adding anything new to the discussion.

**Comment by forthy42** [comment]
No, the comment doesn't.  I've read it, and I'm not convinced.

First of all: If you need the UUIDs to be unique, either use a reliable nonce creating technique, or do something else to avoid collisions (e.g. re-get another one, and fall back to increment if that fails, too).  You said you don't want to maintain your own PRNG, but a cryptographic random number generator definitely can't promise you to deliver unique numbers (it's just pretty unlikely that within a small bunch of 64 bit numbers, two are equal — but if you boot on billions of computers, and create many UUIDs, it will probably happen to someone, sometimes).  LCGs the way you use them (lower 24 bits) will give you non-unique results with a much higher likelihood, which is why rand() on older glibcs can't be used, and I actually don't know how good the new random() is for that purpose (it's better, but is it good enough?).  Unless you add the inevitable collision avoidance code, which is what you should have done in the first place, anyways.

/dev/urandom has a PRNG that passes all relevant tests (even if badly seeded), but it will give you the same number of random collisions as any other true random source (few, but not zero).  You'll likely never see them in your lab, and if the user reboots if it fails to boot once, he also won't see it again.  Don't treat a random number as predictable nonce; if you want a nonce, use a deliberate pedictable nonce function.  If you want to be cryptographically hard, the only nonce function you can use is a counter encrypted with a secret key and a block cipher (take the full block then, it will not repeat as long as the counter doesn't repeat).  If you can accept external predictability, just use a counter.  If you can detect collisions, add avoidance code.  Since systemd fails with a lot of error messages, you obviously can detect collisions, you then can also avoid them.

About trusting CPUs: We know that we can't trust RdRand, because it's such an easy attack vector.  Nobody trusts it.  AMD got away with their bug in their own lab, because nobody trusts it (apart from recent systemd, which trusts it even to deliver a nonce, which was never part of the promise).  Don't trust it.  We are not fucked up if we know where we can't trust the CPU.  We are not fucked up if we can't trust the entropy, but then, we need to test the entropy before we use it.

Actually, cryptographic hard code does a lot of stuff to avoid those parts of CPUs where we know we can't trust them.  We need particular code to minimize side channel leaks of secrets (conditional moves instead of branches, no memory indexed by secrets, etc.; side channel leaks are more than a decade older than Spectre), we need to work around over-optimization in C compilers to make sure our secret copies are erased after use (there was an hour-long talk on 35c3 about only that topic, a conceptual simple thing like predictable erase memory is already hard — because compiler writer don't understand that this is not something that can be optimized away).  We are not fucked up if things don't quite work as they should.

**Comment by forthy42** [comment]
BTW, one comment to improve the auditability of this code (it's going to be looked at by many people, anyhow):

This random generator has a number of different users (callers) that have clearly very different requirements.  It satisfies those by having a bunch of flags, and different code paths for the different flags; part of the flags are obvious optimizations, i.e. for speed.  This makes it pretty hard to audit.

It would be much cleaner if each of the different requirements had their own function including with spec documented, e.g. get_random_uuid(uuid * x, int (*check_collision)(uuid* x)) for the UUID stuff.  The spec should say what you need, like

[x] unique
[x] with check function to verify that
[x] unpredictable
[ ] fast
[x] reliable (caller has no fallback)
[64] bits

something like that.  Then auditors can look at it and quickly know what a particular function needs to do.

Do one thing and do that right is a good idea for functions in C programs, too.

**Comment by TCM2** [comment]
> Do one thing and do that right is a good idea for functions in C programs, too.

Which is precisely why you'll never see it in systemd code.

**Comment by poettering** [comment]
> No, the comment doesn't. I've read it, and I'm not convinced.
> 
> First of all: If you need the UUIDs to be unique, either use a reliable nonce creating technique, or do something else to avoid collisions (e.g. re-get another one, and fall back to increment if that fails, too). You said you don't want to maintain your own PRNG, but a cryptographic random number generator definitely can't promise you to deliver unique numbers (it's just pretty unlikely that within a small bunch of 64 bit numbers, two are equal — but if you boot on billions of computers, and create many UUIDs, it will probably happen to someone, sometimes). 

Well, it's how uuids are generated these days. Pretty much everyone just uses type 4 uuids (i.e. randomized ones). If you think that the likeliness of collision is too high, by all means, fix the whole stack, but please start with the linux kernel, which for example generates uuids this way via /proc/sys/kernel/random/uuid or /proc/sys/kernel/random/boot_id.

I am pretty sure we are fine with this mode of operation as long as the Linux kernel is.

And no, the other ways to generate uuids don#t really work: maintaining global counters is either unsafe (because we need uuids in all kinds of processes, and we can't share state safely if they run with different privileges and shall not be able to trigger collisions easily). And things like uuidd doesn't work during early boot either.

> LCGs the way you use them (lower 24 bits) will give you non-unique results with a much higher likelihood, which is why rand() on older glibcs can't be used, and I actually don't know how good the new random() is for that purpose (it's better, but is it good enough?). Unless you add the inevitable collision avoidance code, which is what you should have done in the first place, anyways.

We don't use glibc rand() for generating uuids, and that's expressly documented in our comments.

> About trusting CPUs: We know that we can't trust RdRand, because it's such an easy attack vector. Nobody trusts it. AMD got away with their bug in their own lab, because nobody trusts it (apart from recent systemd, which trusts it even to deliver a nonce, which was never part of the promise). Don't trust it. We are not fucked up if we know where we can't trust the CPU. We are not fucked up if we can't trust the entropy, but then, we need to test the entropy before we use it.

Yes, but already discussed to length: Thing though is that a) we don't trust RDRAND for generating cryptographic keys. When we do that we go to /dev/urandom like everyone should. b) we trust it enough for generating uuids, which are not cryptographic key material, where RDRAND should be fine and c) The linux kernel on various prominent distros actually does trust RDRAND now, and will fill /dev/urandom with it crediting it for full entropy, so that we'll actually end up using RDRAND after all in the end... i.e. check this:

```
$ grep CONFIG_RANDOM_TRUST_CPU /usr/lib/modules/`uname -r`/config
CONFIG_RANDOM_TRUST_CPU=y
```
(that's from Fedora)

So, yes, in systemd we are not going to make the decision to rely on RDRAND for key generation, but your distro might have configured to /dev/urandom like that, so we might ultimately after all.

systemd uses RDRAND directly only for generating UUIDs, seeding hash tables, temporary file name starting seeds, unix UID acquisition statring seeds, where things generally don't matter that much.

> Actually, cryptographic hard code does a lot of stuff to avoid those parts of CPUs where we know we can't trust them. We need particular code to minimize side channel leaks of secrets (conditional moves instead of branches, no memory indexed by secrets, etc.; side channel leaks are more than a decade older than Spectre), we need to work around over-optimization in C compilers to make sure our secret copies are erased after use (there was an hour-long talk on 35c3 about only that topic, a conceptual simple thing like predictable erase memory is already hard — because compiler writer don't understand that this is not something that can be optimized away). We are not fucked up if things don't quite work as they should.

Yeah, sure, for cryptographic stuff we use /dev/urandom, read the sources and its comments, it makes that clear. 

> BTW, one comment to improve the auditability of this code (it's going to be looked at by many people, anyhow):
> 
> This random generator has a number of different users (callers) that have clearly very different requirements. It satisfies those by having a bunch of flags, and different code paths for the different flags; part of the flags are obvious optimizations, i.e. for speed. This makes it pretty hard to audit.
> 
> It would be much cleaner if each of the different requirements had their own function including with spec documented, e.g. get_random_uuid(uuid * x, int (_check_collision)(uuid_ x)) for the UUID stuff. The spec should say what you need, like

uh? callback functions? what's that supposed to do? also, anything with such callback functions probably makes things harder to read, not simpler. (I know, because I spend 90% of my days reviewing code, do you?)

**Comment by poettering** [comment]
BTW, AMD is preparing a firmware update now that fixes RDRAND for them (not sure of just one new Ryzen or on old Ryzen too):

https://www.heise.de/newsticker/meldung/AMD-beseitigt-Fehler-des-Ryzen-3000-ueber-BIOS-Updates-4468815.html

(sorry, German only)

**Comment by forthy42** [comment]
The Linux kernel uses a (mostly) sane construction: It seeds an entropy pool and uses a cryptographic hard (chacha20-based) PRNG to emit a stream of randomness. With 16 bytes out of a chacha20-based PRNG with a central entropy pool (so no way to read the same value twice; there's a central counter besides the entropy pool), collision is clearly low enough to be tolerable.

Remaining sanity problems: The kernel uses rdrand() in some places (get_random_u64(), e.g. used as random ID for seccomp stuff; this is probably arcane enough).  What it should do is using get_random_bytes() that does not block when there's not enough entropy in the pool.  This still delivers a stream of cryptographic hard stream of randomness that is to some extend predictable (not enough entropy in the pool), but certainly won't provide collisions.

This is something the Linux kernel only partly delivers to userland, but is worth having.  Implementation would be straight-forward, just another flag to getrandom() which returns a bunch of pseudo-random bytes without blocking, even when the pool is not yet filled with trustworthy entropy, but only with the non-trustworthy RDSEED entropy.  This would relieve the systemd's problem that it needs entropy very early in the boot sequence.

Essentially, it's a one-line patch of random.c, line 2135:

-if (!crng_ready()) {
+if (!crng_ready() && !(flags & GRND_PREMATURE) {

BTW: Directly reading from /dev/urandom has the same effect, as unlike getrandom() it does never block, only warn when there's no entropy.

The point is that when you want collision-free stuff, you have about four options:

a) use a known cryptographic hard random source that has central entropy pool+state (entropy doesn't matter much, it matters that reads always give non-overlapping parts of the output stream)
b) a nonce generator that is guaranteed to deliver non-repeating numbers (which, BTW, even the LCG rand() does, as long as it follows the central entropy pool+state paradigm which is probably difficult if that are different processes, and the seed may fail); cryptographic hard nonce generators are e.g. block ciphers in counter mode with a random seed as key
c) a central counter (not worrying about predictability)
d) a entropy source that is good enough to pass the randomness test suite like dieharder (looks like AMD's entropy source actually is that, because RDRAND and RDSEED fail the same way — that indicates that they don't have a conditioning PRBG in RDRAND, just an entropy source with an inadequate power saving mode, and RDRAND+RDSEED are identical)

If you want to scale, and not rely on a centralized pool or state, you can use construct a) with per-process/thread PRBG (pseudo-random; if you a block cipher with a counter as tweak, a per-thread guaranteed non-overlapping starting point of the counter is sufficient); the seed for that can come from something like randombytes(), or you use construct d) with per-core entropy sources (non-deterministic NRBG).

The thing that is no option is a entropy source that is not either monitored for quality or by audit known good.  The failure modes of hardware are not necessarily the ones you think they are.  “Don't matter that much” assumes maybe a well-hidden backdoor (indistinguishable from randomness unless you have the key), not a rather hard failure mode, e.g. entropy source not powered up.

You don't trust RDRAND for cryptography, but you do trust it for non-collision.  That works for the Intel case, where RDRAND is the output of a stream cipher (AES in counter mode with the key coming from the entropy source — if the entropy source fails, the key stream still will look indistinguishable from proper random).  That doesn't work for the AMD case, where a failing entropy source also causes RDRAND to fail the same way as RDSEED.

A dubious quality random number source where you can know when it is failing (by simply analyzing it) is probably one of the better deals you can get in this situation.  It helps if you can only destroy it, but can't make it predictable by someone who knows a secret.

BTW: You missed the main point in the last part: have separate functions for separate purposes, and clearly specify what they should deliver.  The callback for checking for collisions was an example.  The main thing is that you properly specify what you need, and then you can evaluate whether you call the right function.  Linux has generate_random_uuid(), which calls get_random_bytes(), which warns only if the entropy is insufficient, but delivers a collision-unlikely PRNG output nonetheless.

This is something that is quite straight-forward to audit, even when it lacks comments.  Anyways: with the exception of two or three cases where get_random_u64 is used in the Linux kernel, and the lack of a pseudo-random fallback mode for getrandom(), there's almost nothing to complain there: This uses a state of the art construction with entropy source+predictable expansion.  The code there has clearly been written by someone who knows what he was doing, and has changed completely after 2013 (for good).

**Comment by owlstead** [comment]
> You don't trust RDRAND for cryptography, but you do trust it for non-collision. That works for the Intel case, where RDRAND is the output of a stream cipher (AES in counter mode with the key coming from the entropy source — if the entropy source fails, the key stream still will look indistinguishable from proper random). That doesn't work for the AMD case, where a failing entropy source also causes RDRAND to fail the same way as RDSEED.

Of course this doesn't work for Intel either. The key stream will look random *on it's own* but it would be indistinguishable from an earlier generated stream that uses the same seed material. Maybe that won't wrong-foot the mentioned systemd, but it would break any other system that relies on it to not collide with a previous stream. For instance, let's assume that you start by generating UUIDs: these are supposed to be unique across systems, not just the booting system.

**Comment by tytso** [comment]
So that's not quite rigth.  RDRAND is fine for creating random cryptographic keys; it is NIST SP 800-90A complaint.   However, if you need to seed another cryptographic generator, you should use RDSEED.   For more details, see [1]

[1] https://software.intel.com/en-us/blogs/2012/11/17/the-difference-between-rdrand-and-rdseed

All of this is of course if you trust Intel, of course.   Of course, I'd probably trust Intel much more than I would trust an ARM chip manufactured by Huawei --- but someone working for the Ministry of State Security or the Chinese People's Liberation Army might have an opposite point of view on that score.

As far as GRND_PREMATURE is concerned, it is something I've considered, although it would probably be something like GRND_I_SOLEMNLY_SWEAR_I_UNDERSTAND_THIS_IS_INSECURE.   Ultimately, it's about how much do you trust application programmers, and whether someone (perhaps a less sophisticated distribution packager/maintainer) might blindly apply a patch that adds GRND_PREMATURE to "fix" a hang during early boot problem, when it's being used to create long-term public/private keypairs.   One could argue that silly/stupid programmers like that might very well use /dev/urandom, and that's a fair point.   My preference would be to simply tell people if you really need something which is unique, but not necessarily cryptographically secure, to simply read a UUID from /proc/sys/kernel/random/uuid.   Yes, today it will potentially issue a warning to dmesg since it calls get_random_bytes(), but I'd much rather fix the kernel to suppress a warning if someone fetches a UUID via /proc/sys/kernel/random/uuid than to add a new GRND_WATCH_ME_GET_ABUSED_BY_A_CLUELESS APPLICATION_PROGRAMMER flag.   Because I can guarantee it will happen, and probably it will get backed into some trashy IOT device....

**Comment by poettering** [comment]
hmmm, but with a totally not initialized pool on a system that is entropy starved (i.e. no rdrand, no hwrng of any form), generating a uuid from  /proc/sys/kernel/random/uuid wouldn't be solving our problem either, would it? i mean, if there's really no entropy than pulling a uuid from that would result in basically a constant value, which defeats the purpose of uuids...

**Comment by tytso** [comment]
It'll be at least as good as using /dev/urandom, and you're OK with that, right?    We will seed the entropy pool with the hardware TOD clock, and with the ethernet addresses for any device drivers that are loaded (although if they are all modules, that'll be an ordering issue), etc.    So if you're going for uniqueness but not necessarily cryptographically secure, in most cases it will hopefully be good enough.     Now if the there is no hardware TOD clock (no batteries, and you're utterly reliant on ntp), and all of your networking devices are modules loaded very late in the boot process, this might no be all that helpful --- but then you're screwed if you use /dev/urandom.  Again, /proc/sys/kernel/random/uuid is no worse than /dev/urandom.   And today it *will* issue a kernel warnings, but that is something I'll suppress in the future, especially if we can advertise that the only safe things people should be using is /proc/sys/kernel/random/uuid or getrandom(2), and there may be times when the uuid won't be unique, but ultimately, if you're using crap designed hardware, we can't manufacture uniqueness (or randomness) where there is none.....

**Comment by forthy42** [comment]
Am 24. Juli 2019 01:57:54 GMT+08:00 schrieb tytso <notifications@github.com>:
>So that's not quite rigth.  RDRAND is fine for creating random
>cryptographic keys; it is NIST SP 800-90A complaint.

The assumption was that the entropy source fails. 

>However, if you
>need to seed another cryptographic generator, you should use RDSEED.  
>For more details, see [1]
>
>[1]
>https://software.intel.com/en-us/blogs/2012/11/17/the-difference-between-rdrand-and-rdseed
>
>All of this is of course if you trust Intel, of course.   Of course,
>I'd probably trust Intel much more than I would trust an ARM chip
>manufactured by Huawei --- but someone working for the Ministry of
>State Security or the Chinese People's Liberation Army might have an
>opposite point of view on that score.

The PLA would have performed an audit of the Huawei device, and can't do the same with Intel. OK, maybe they can (by hacking into Intel's EDA computers ;-)…

>As far as GRND_PREMATURE is concerned, it is something I've considered,
>although it would probably be something like
>GRND_I_SOLEMNLY_SWEAR_I_UNDERSTAND_THIS_IS_INSECURE.

Haha, fine with me. I thought about GRND_TRUST_CPU_ENTROPY_SOURCE, similar to the kernel compilation switch. Or GRND_JUST_GIVE_ME_UUIDS_QUICK_PLEASE

> Ultimately, it's
>about how much do you trust application programmers, and whether
>someone (perhaps a less sophisticated distribution packager/maintainer)
>might blindly apply a patch that adds GRND_PREMATURE to "fix" a hang
>during early boot problem, when it's being used to create long-term
>public/private keypairs.

I'd expect them to use the nonblock option, and ignore the error code, maybe filling the memory block with time+MAC before. People resort to desperate means when they need to get a crappy product out into production tomorrow. Filling in the requested memory with the chacha20 stream anyways for nonblockig requests could be a good idea; the returned EAGAIN then means that it is not seeded properly. 

>One could argue that silly/stupid
>programmers like that might very well use /dev/urandom, and that's a
>fair point.   My preference would be to simply tell people if you
>really need something which is unique, but not necessarily
>cryptographically secure, to simply read a UUID from
>/proc/sys/kernel/random/uuid.   Yes, today it will potentially issue a
>warning to dmesg since it calls get_random_bytes(), but I'd much rather
>fix the kernel to suppress a warning if someone fetches a UUID via
>/proc/sys/kernel/random/uuid than to add a new
>GRND_WATCH_ME_GET_ABUSED_BY_A_CLUELESS APPLICATION_PROGRAMMER flag.  
>Because I can guarantee it will happen, and probably it will get backed
>into some trashy IOT device....

If the purpose is getting an UUID, it's OK. File interfaces are problematic, as they might not be mounted (e.g. early in the boot sequence, or in a jailed environment). That was the original point of providing a kernel call for this function. 

-- 
Bernd Paysan
"If you want it done right, you have to do it yourself"
net2o ID: kQusJzA;7*?t=uy@X}1GWr!+0qqp_Cn176t4(dQ*
http://bernd-paysan.de/

**Comment by poettering** [comment]
> It'll be at least as good as using /dev/urandom, and you're OK with that, right? 

No, we do not use /dev/urandom for UUID generation (at least not on systems that support getrandom() that is, on older stuff we do, simply because we don't have anything better and then use /dev/urandom as replacement for getrandom()). On modern systems, for UUID generation we'll try RDRAND first, and synchronous getrandom(flags=0) second, and that's where the story ends.

> We will seed the entropy pool with the hardware TOD clock, and with the ethernet addresses for any device drivers that are loaded (although if they are all modules, that'll be an ordering issue), etc. So if you're going for uniqueness but not necessarily cryptographically secure, in most cases it will hopefully be good enough.

Typical general purpose systems are usually not linking ethernet drivers in, which leaves the clock as only source, but if you start multiple systems at once (which is likely enough in virtualized environments) there's a good chance they then all generate the same series of uuids, which we don't want...

>  Now if the there is no hardware TOD clock (no batteries, and you're utterly reliant on ntp), and all of your networking devices are modules loaded very late in the boot process, this might no be all that helpful --- but then you're screwed if you use /dev/urandom. Again, /proc/sys/kernel/random/uuid is no worse than /dev/urandom. And today it _will_ issue a kernel warnings, but that is something I'll suppress in the future, especially if we can advertise that the only safe things people should be using is /proc/sys/kernel/random/uuid or getrandom(2), and there may be times when the uuid won't be unique, but ultimately, if you're using crap designed hardware, we can't manufacture uniqueness (or randomness) where there is none.....

I would prefer if the warning would stay for /proc/sys/kernel/random/uuid too... In virtualized environments the chance is too high that on general purpose distros this will generate a non-unique stream of uuids.

(BTW, related to this, in #13137 I prepped a PR that makes systemd credit entropy with data stored in a specific EFI variable to the kernel pool if its set, very early on. It also contains work that systemd-boot (which is an EFI boot loader included in our tree) reads a random seed file from the ESP, then generates two new seeds from it with SHA256 counter mode, rewriting the random seed file with one, and writing the other into the EFI variable. This should then give us a full pool from earliest userspace on, at least on EFI systems that use systemd-boot, and if the random seed file in the ESP was initialized with good entropy first. It takes inspiration from NetBSD's boot loader random seed stuff, but is a bit more careful to never reuse the same unmodified seed in multiple boots on the same or on different systems. Ideally the kernel would already read that EFI variable itself btw, and unset it. And ideally there'd be a non-EFI mechanism for this too.)

**Comment by poettering** [comment]
> If the purpose is getting an UUID, it's OK. File interfaces are problematic, as they might not be mounted (e.g. early in the boot sequence, or in a jailed environment). That was the original point of providing a kernel call for this function.

Nah, systemd runs as PID 1, it always mounts /proc as first thing, and the first uuid it needs afterwards. We are totally OK with use /proc if we need to. That said, we don't want to generate uuids from an uninitialized pool, since the chance is too high for it not resulting in unique ids.

**Comment by poettering** [comment]
Just to mention this: it appears AMD admits to the borkage now, and now posted this:

https://lore.kernel.org/patchwork/patch/1115413/

In times of most big distros defaulting to CONFIG_RANDOM_TRUST_CPU=y this sounds like a big CVE-worthy security vuln, but apparently noone cares enough for AMD CPUs?

Also, I put together this:

https://systemd.io/RANDOM_SEEDS 

That document explains systemd logic behind random seed initialization and RDRAND usage.

**Comment by noloader** [comment]
In case someone is interested in picking up a test rig... The [dell inspiron 3180](https://www.google.com/search?q=dell+inspiron+3180+used) is a 15h machine (cpu_family = 21) and should have the buggy instruction. A used one costs about $100 USD.

**Comment by noloader** [comment]
AMD provides a secure RNG library at [RNG Library](https://developer.amd.com/amd-aocl/rng-library/). It was last updated Jan 15, 2020.

The latest secure RNG code is available in `aocl-securerng-linux-aocc-2.1.tar.gz`. The release notes and source code do not discuss the RDRAND bug:

```
amd-securerng$ egrep -iIR 'bug|resume'
...
ReleaseNotes_AMDSecureRNG.txt:-   Bug fixes and improvements
amd-securerng$
```

Looking at the source code in `src_lib/secure.c`, it appears the library just uses `_rdrand32_step` (and friends), and checks the return value of the intrinsic. The return value is set by carry flag from the instruction.

It is not clear (to me) if all that is needed is to check the carry flag.

However, the release notes skip 15h and 16h processors, and moves on to 17h processors:

```
System Requirements
-------------------

Hardware - Processor should support RDRAND and RDSEED instructions. AMD Family 
17h processors and other products of 2016 and beyond support these instructions
```

It is a pretty shitty way to do things... don't discuss the bug, and don't tell folks how to use the instruction safely.

**Comment by poettering** [comment]
> Looking at the source code in `src_lib/secrng.c`, it appears the library just uses `_rdrand32_step` (and friends), and checks the return value of the intrinsic. The return value is set by carry flag from the instruction.
> 
> It is not clear (to me) if all that is needed is to check the carry flag.

We check the carry flag. Checking the carry flag is documented, and hence we do it. However, on the affected AMD CPU flag they set the carry flag incorrectly and hand us constant data with no indication that it is rubbish

**Comment by DwarfFighterCleric** [comment]
AMD/Ryzen users: was this ever resolved?  
I have been unable to Suspend my brand new Ryzen laptop to RAM, using latest-and-greatest Nov2020 ISO's of both Manjaro and MX-Linux (Advanced Hardware Support AHS 19.3 respin also tested to be sure) Fedora 33 tested as well.  
Nov2020, still going on.

**Comment by sandy-8925** [comment]
> AMD/Ryzen users: was this ever resolved?
> I have been unable to Suspend my brand new Ryzen laptop to RAM, using latest-and-greatest Nov2020 ISO's of both Manjaro and MX-Linux (Advanced Hardware Support AHS 19.3 respin also tested to be sure) Fedora 33 tested as well.
> Nov2020, still going on.
> ... (only AMD users to answer, if possible, since you are the ones feeling this in the flesh)
> Question 2: Is this issue only present on laptops or desktops as well? Suspend to RAM

Your problem most likely has nothing to do with RDRAND, you should debug and find out why exactly it's happening.

I believe systemd did create a workaround to check the output of RDRAND and discard it if it's buggy. AMD did also release corrected AGESA code, and a few motherboard manufacturers have released UEFI updates that include the updated AGESA. You should make sure to have the latest UEFI updates installed, and confirm with your motherboard manufacturer that they have released a fix with the relevant or newer AGESA version.

**Comment by DwarfFighterCleric** [comment]
My guy, I don't even know what is RDRAND, I'm just a "regular user" that happens to own AMD Ryzen APU.  
I just noticed there was a bunch of AMD users here, so I thought "Looks like a nice place to ask questions to fellow AMD owners and AMD regular users"  
I'm just asking this question (this is not a bug report or request):  
Do you use AMD Ryzen? If Yes, are you able to suspend your Linux OS to RAM, like worked on older versions?  
That's all.  
I just want to know if this was ever resolved (things preventing suspend to work) because from what I've read in other places this was never properly addressed. The only thing that happened is a bunch of people saying things like "I never suspend to ram anyway, just suspend to disk instead" I'm not interested in cop-outs.

**Comment by sandy-8925** [comment]
> My guy, I don't even know what is RDRAND, I'm just a "regular user" that happens to own AMD Ryzen APU.
> I just noticed there was a bunch of AMD users here, so I thought "Looks like a nice place to ask questions to fellow AMD owners and AMD regular users"
> I'm just asking this question (this is not a bug report or request):
> Do you use AMD Ryzen? If Yes, are you able to suspend your Linux OS to RAM, like worked on older versions?
> That's all.
> I just want to know if this was ever resolved (things preventing suspend to work) because from what I've read in other places this was never properly addressed. The only thing that happened is a bunch of people saying things like "I never suspend to ram anyway, just suspend to disk instead" I'm not interested in cop-outs.

Well in that case, this is not the appropriate place to ask about suspend issues, since it's not RDRAND related. What you need to do is file a bug report in your distro's bug tracker and go from there. And maybe ask for help on some tech support forums.

**Comment by dj-on-github** [comment]
If you need to know about RdRand, ask me.

**Comment by keszybz** [comment]
Just not here, if you may.
